pyigm/fN/fnmodel.py

```diff
@@ -0,0 +1,456 @@
+            lX2 = np.zeros(nz)
+            for ii in range(nz):
+                lgfNX[:, ii] = self.evaluate(lgNHI2, z[ii]).flatten()
+                lX2[ii] = np.sum(10.**(lgfNX[:, ii]+lgNHI2)) * dlgN * np.log(10.)
+            lX = lX + lX2
+
+        # Return
+        if nz == 1:
+            lX = lX[0]
+        if cumul == True:
+            return lX, cum_sum, lgNHI
+        else:
+            return lX
+
+    def calculate_rhoHI(self, z, NHI_mnx, neval=10000, cumul=False,
+        H0=70.*u.km/(u.s*u.Mpc)):
+        """ Calculate rho_HI over an N_HI interval
+
+        Parameters
+        ----------
+        z : float
+          Redshift for evaluation
+        NHI_mnx : tuple
+          of floats minimum/maximum NHI values
+        neval : int, optional
+          Discretization parameter (10000)
+        cumul : bool, optional
+          Return a cumulative array?
+        H0 : float, optional
+          Hubble's constant with units [70 km/s/Mpc]
+
+        Returns
+        -------
+        rho_HI: float
+          rho_HI in units of Msun per comoving Mpc**3
+        """
+        # Initial 
+        NHI_min=NHI_mnx[0]
+        NHI_max=NHI_mnx[1]
+        try:
+            nz = len(z)
+        except:
+            nz=1
+            z = np.array([z])
+
+        # Brute force (should be good to ~0.5%)
+        lgNHI = NHI_min + (NHI_max-NHI_min)*np.arange(neval)/(neval-1.)
+        dlgN = lgNHI[1]-lgNHI[0]
+
+        # Evaluate f(N,X)
+        lgfNX = self.evaluate(lgNHI, z)
+
+        # Sum
+        rho_HI = np.zeros(nz)
+        for ii in range(nz): 
+            rho_HI[ii] = np.sum(10.**(lgfNX[:, ii]+2*lgNHI)) * dlgN * np.log(10.)
+        if cumul==True: 
+            if nz > 1:  #; Have not modified this yet
+                raise ValueError('fN.model: Not ready for this model type %s' % self.mtype)
+            cum_sum = np.cumsum(10.**(lgfNX[:, ii]+2*lgNHI)) * dlgN * np.log(10.)
+
+        # Constants
+        rho_HI = rho_HI * (const.m_p.cgs * H0 / 
+            const.c.cgs / (u.cm**2)).to(u.Msun/u.Mpc**3)
+
+        # Return
+        if nz == 1:
+            rho_HI = rho_HI[0]
+        if cumul == True:
+            return rho_HI, cum_sum, lgNHI
+        else:
+            return rho_HI
+
+    # Evaluate
+    def evaluate(self, NHI, z, vel_array=None, cosmo=None):
+        """ Evaluate the f(N,X) model at a set of NHI values
+
+        Parameters
+        ----------
+
+        NHI : array
+          NHI values
+        z : float or array
+          Redshift for evaluation
+        vel_array : ndarray, optional
+          Velocities relative to z
+        cosmo : astropy.cosmology, optional
+
+
+        Returns
+        -------
+
+        log_fNX : float, array, or 2D array
+          f(NHI,X)[z] values
+          Float if given one NHI,z value each. Otherwise 2D array
+          If 2D, it is [NHI,z] on the axes
+
+        """
+        # Tuple?
+        if isinstance(NHI, tuple): # All values packed into NHI parameter
```

Comment:
Agreed.  And I forget why I was forced to do this contortion.
Could try to undo it if you greatly prefer.